app/web_conn.go

```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
 	}
 }
 
-// writeMessage is a helper utility that wraps the write to the socket
+// writeMessageBuf is a helper utility that wraps the write to the socket
 // along with setting the write deadline.
-func (wc *WebConn) writeMessage(msgType int, data []byte) error {
+func (wc *WebConn) writeMessageBuf(msgType int, data []byte) error {
 	wc.WebSocket.SetWriteDeadline(time.Now().Add(writeWaitTime))
 	return wc.WebSocket.WriteMessage(msgType, data)
 }
 
+func (wc *WebConn) writeMessage(msg *model.WebSocketEvent) error {
+	// We don't use the encoder from the write pump because it's unwieldy to pass encoders
+	// around, and this is only called during initialization of the webConn.
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	err := msg.Encode(enc)
+	if err != nil {
+		mlog.Warn("Error in encoding websocket message", mlog.Err(err))
+		return nil
+	}
+	wc.Sequence++
+
+	return wc.writeMessageBuf(websocket.TextMessage, buf.Bytes())
+}
+
 // addToDeadQueue appends a message to the dead queue.
-func (wc *WebConn) addToDeadQueue(msg model.WebSocketMessage) {
+func (wc *WebConn) addToDeadQueue(msg *model.WebSocketEvent) {
 	wc.deadQueue[wc.deadQueuePointer] = msg
 	wc.deadQueuePointer = (wc.deadQueuePointer + 1) % deadQueueSize
 }
 
+// hasMsgLoss indicates whether the next wanted sequence is right after
+// the latest element in the dead queue, which would mean there is no message loss.
+func (wc *WebConn) hasMsgLoss() bool {
+	if wc.deadQueuePointer == 0 {
+		if wc.deadQueue[deadQueueSize-1] == nil {
+			return false // No msg written
+		}
+		if wc.deadQueue[deadQueueSize-1].GetSequence() == wc.Sequence-1 {
+			return false
+		}
+		return true
+	}
+	if wc.deadQueue[wc.deadQueuePointer-1].GetSequence() == wc.Sequence-1 {
+		return false
+	}
+	return true
+}
```

Comment:
Interestingly I find this logic a bit hard to digest and I think it's mainly because we are doing two things at a time and returning in so many places. I am wondering if it would be clearer if we first find the element we are looking for (the last) and then do the sequence check (if not nil of course).

Reply:
Ah ok, let me see what I can do.